File: src/main/scala/resonant/lib/grid/UpdateTicker.java
package resonant.lib.grid;

import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Queue;
import java.util.Set;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.Event;
import universalelectricity.api.net.IUpdate;

/** A ticker to update all grids. This is multithreaded.
 * 
 * @author Calclavia */
public class UpdateTicker extends Thread
{
    public static final UpdateTicker INSTANCE = new UpdateTicker();

    /** For updaters to be ticked. */
    private final Set<IUpdate> updaters = Collections.newSetFromMap(new WeakHashMap<IUpdate, Boolean>());

    /** For queuing Forge events to be invoked the next tick. */
    private final Queue<Event> queuedEvents = new ConcurrentLinkedQueue<Event>();

    public boolean pause = false;

    /** The time in milliseconds between successive updates. */
    private long deltaTime;

    public UpdateTicker()
    {
        setName("Universal Electricity");
        setPriority(MIN_PRIORITY);
    }

    public static void addNetwork(IUpdate updater)
    {
        synchronized (INSTANCE.updaters)
        {
            INSTANCE.updaters.add(updater);
        }
    }

    public static synchronized void queueEvent(Event event)
    {
        synchronized (INSTANCE.queuedEvents)
        {
            INSTANCE.queuedEvents.add(event);
        }
    }

    public long getDeltaTime()
    {
        return deltaTime;
    }

    public int getUpdaterCount()
    {
        return updaters.size();
    }

    @Override
    public void run()
    {
        try
        {
            long last = System.currentTimeMillis();

            while (true)
            {
                if (!pause)
                {
                    // theProfiler.profilingEnabled = true;
                    long current = System.currentTimeMillis();
                    deltaTime = current - last;

                    /** Tick all updaters. */
                    synchronized (updaters)
                    {
                        Set<IUpdate> removeUpdaters = Collections.newSetFromMap(new WeakHashMap<IUpdate, Boolean>());

                        Iterator<IUpdate> updaterIt = new HashSet<IUpdate>(updaters).iterator();

                        try
                        {
                            while (updaterIt.hasNext())
                            {
                                IUpdate updater = updaterIt.next();

                                if (updater.canUpdate())
                                {
                                    updater.update();
                                }

                                if (!updater.continueUpdate())
                                {
                                    removeUpdaters.add(updater);
                                }
                            }

                            updaters.removeAll(removeUpdaters);
                        }
                        catch (Exception e)
                        {
                            System.out.println("Universal Electricity Threaded Ticker: Failed while tcking updater. This is a bug! Clearing all tickers for self repair.");
                            updaters.clear();
                            e.printStackTrace();
                        }
                    }

                    /** Perform all queued events */
                    synchronized (queuedEvents)
                    {
                        while (!queuedEvents.isEmpty())
                        {
                            MinecraftForge.EVENT_BUS.post(queuedEvents.poll());
                        }
                    }

                    last = current;
                }

                Thread.sleep(50L);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: src/main/scala/resonant/lib/grid/UpdateTicker.scala
package resonant.lib.grid

import java.util
import java.util.Collections
import java.util.concurrent.ConcurrentLinkedQueue

import cpw.mods.fml.common.eventhandler.{Event, SubscribeEvent}
import cpw.mods.fml.common.gameevent.TickEvent
import cpw.mods.fml.common.gameevent.TickEvent.Phase
import net.minecraftforge.common.MinecraftForge
import resonant.api.IUpdate

import scala.collection.convert.wrapAll._

/**
 * A ticker to update all grids. This is multi-threaded based on configuration.
 *
 * @author Calclavia
 */
object UpdateTicker
{
  final val threaded = new UpdateTicker
  final val world = new UpdateTicker
}

class UpdateTicker extends Thread
{
  setName("Universal Electricity")
  setPriority(Thread.MIN_PRIORITY)

  /**
   * For updaters to be ticked.
   */
  private final val updaters: java.util.Set[IUpdate] = Collections.newSetFromMap(new util.WeakHashMap[IUpdate, java.lang.Boolean]())

  private final val queue = new ConcurrentLinkedQueue[() => Unit]()

  /**
   * For queuing Forge events to be invoked the next tick.
   */
  private final val queuedEvents = new ConcurrentLinkedQueue[Event]()

  /**
   * Becomes true if the network needs to be paused.
   */
  var pause = false

  /**
   * The time in milliseconds between successive updates.
   */
  private var deltaTime = 0L

  def addUpdater(updater: IUpdate)
  {
    enqueue(() => updaters.add(updater))
  }

  def enqueue(f: () => Unit)
  {
    queue.add(f)
  }

  def queueEvent(event: Event)
  {
    queuedEvents.add(event)
  }

  def getDeltaTime = deltaTime

  def getUpdaterCount = updaters.size

  override def run()
  {
    var last = System.currentTimeMillis()

    while (true)
    {
      if (!pause)
      {
        val current = System.currentTimeMillis()
        deltaTime = current - last
        update()

        last = current
      }

      Thread.sleep(50L)
    }
  }

  @SubscribeEvent
  def tickEnd(event: TickEvent.ServerTickEvent)
  {
    if(event.phase == Phase.END)
    {
      deltaTime = 50
      update()
    }
  }

  def update()
  {
    try
    {
      queue synchronized
      {
        queue.foreach(_.apply())
        queue.clear()
      }

      updaters synchronized
      {
        if (this == UpdateTicker.threaded)
          updaters.par.filter(_.canUpdate()).foreach(_.update(getDeltaTime / 1000d))
        else
          updaters.filter(_.canUpdate()).foreach(_.update(getDeltaTime / 1000d))

        updaters.removeAll(updaters.filterNot(_.continueUpdate()))
      }

      queuedEvents synchronized
      {
        queuedEvents.foreach(MinecraftForge.EVENT_BUS.post)
        queuedEvents.clear()
      }
    }
    catch
      {
        case e: Exception =>
        {
          System.out.println("Universal Electricity Ticker: Failed while ticking updaters. This is a bug! Clearing all tickers for self repair.")
          updaters.clear()
          e.printStackTrace()
        }
      }
  }
}
